File: docs/vsto/codesnippet/CSharp/trin_vstcorehostcontrolsword/ThisDocument.cs
using System;
using System.Data;
using System.Drawing;
using System.Windows.Forms;
using Microsoft.VisualStudio.Tools.Applications.Runtime;
using Word = Microsoft.Office.Interop.Word;
using Office = Microsoft.Office.Core;

namespace Trin_VstcoreHostControlsWord
{
    public partial class ThisDocument
    {
        //---------------------------------------------------------------------
        //<Snippet4>
        private Office.CommandBar commandBar;
        private Office.CommandBarButton boldText;
        private Office.CommandBarButton italicText;
        private Microsoft.Office.Tools.Word.Bookmark selectedBookmark;

        const int WordTrue = -1;
        const int WordFalse = 0;
        //</Snippet4>


        //---------------------------------------------------------------------
        //<Snippet6>
        private void ThisDocument_Startup(object sender, System.EventArgs e)
        {
            AddPopUp();
        }
        //</Snippet6>


        //---------------------------------------------------------------------
        private void StartUp()
        {
            //<Snippet7>
            this.bookmark1.BeforeRightClick +=
                new Microsoft.Office.Tools.Word.ClickEventHandler(ShowPopupMenu);

            this.bookmark2.BeforeRightClick +=
                new Microsoft.Office.Tools.Word.ClickEventHandler(ShowPopupMenu);
            //</Snippet7>
        }


        //---------------------------------------------------------------------
        //<Snippet5>
        private void AddPopUp()
        {
            commandBar = Application.CommandBars.Add("FormatText",
                Office.MsoBarPosition.msoBarPopup, missing, true);

            // Add a button and set the style, caption, icon and tag.
            boldText = (Office.CommandBarButton)commandBar.Controls.Add(
                1);

            boldText.Style = Office.MsoButtonStyle.msoButtonCaption;
            boldText.Caption = "Bold";
            boldText.FaceId = 113;
            boldText.Tag = "0";

            // Add a button and set the style, caption, icon and tag.
            italicText = (Office.CommandBarButton)commandBar.Controls.Add(
                1);

            italicText.Style = Office.MsoButtonStyle.msoButtonCaption;
            italicText.Caption = "Italic";
            italicText.FaceId = 114;
            italicText.Tag = "1";

            // Handle the click events with the ButtonClick procedure.
            boldText.Click +=
                new Microsoft.Office.Core._CommandBarButtonEvents_ClickEventHandler(ButtonClick);

            italicText.Click +=
                new Microsoft.Office.Core._CommandBarButtonEvents_ClickEventHandler(ButtonClick);

            ((Microsoft.Office.Interop.Word.Template)this.AttachedTemplate).Saved = true;
        }
        //</Snippet5>


        //---------------------------------------------------------------------
        //<Snippet8>
        private void ShowPopupMenu(object sender,
            Microsoft.Office.Tools.Word.ClickEventArgs e)
        {
            int startPosition = 0;

            // If bookmarks overlap, get bookmark closest to cursor.
            for (int i = 1; i <= e.Selection.Bookmarks.Count; i++)
            {
                if (e.Selection.Bookmarks[i].Start > startPosition)
                {
                    startPosition = e.Selection.Bookmarks[i].Start;
                }
            }

            // If closest bookmark is the sender, show the popup.
            if (((Microsoft.Office.Tools.Word.Bookmark)sender).Start == startPosition)
            {
                selectedBookmark = (Microsoft.Office.Tools.Word.Bookmark)sender;
                commandBar.ShowPopup();
                e.Cancel = true;
            }
        }
        //</Snippet8>


        //---------------------------------------------------------------------
        //<Snippet9>
        private void ButtonClick(Microsoft.Office.Core.CommandBarButton Ctrl,
            ref bool CancelDefault)
        {
            if (Ctrl.Caption == "Bold")
            {
                if (selectedBookmark.Bold == WordTrue)
                {
                    selectedBookmark.Bold = WordFalse;
                }
                else
                {
                    selectedBookmark.Bold = WordTrue;
                }
            }
            else if (Ctrl.Caption == "Italic")
            {
                if (selectedBookmark.Italic == WordTrue)
                {
                    selectedBookmark.Italic = WordFalse;
                }
                else
                {
                    selectedBookmark.Italic = WordTrue;
                }
            }
        }
        //</Snippet9>


        //---------------------------------------------------------------------
        private void Test1()
        {
            //<Snippet10>
            foreach (object targetControl in this.Controls)
            {
                Microsoft.Office.Tools.Word.Bookmark bookMark =
                    targetControl as Microsoft.Office.Tools.Word.Bookmark;

                if (bookMark != null)
                {
                    // Do some work with the bookmark here.
                }
            }
            //</Snippet10>


            //<Snippet2>
            this.SampleBookmark.Start = this.SampleBookmark.Start - 5;
            //</Snippet2>


            //<Snippet3>
            this.SampleBookmark.End = this.SampleBookmark.End + 5;
            //</Snippet3>


            //<Snippet1>
            Microsoft.Office.Tools.Word.Bookmark firstParagraph;
            firstParagraph = this.Controls.AddBookmark(this.Paragraphs[1].Range,
                "FirstParagraph");
            //</Snippet1>
        }


        //---------------------------------------------------------------------
        private void InternalStartup()
        {
            this.Startup += new System.EventHandler(ThisDocument_Startup);
        }
    }
}

File: docs/vsto/codesnippet/VisualBasic/Trin_VstcoreHostControlsWordVB/ThisDocument.vb
Public Class ThisDocument

    '--------------------------------------------------------------------------
    '<Snippet4>
    Private commandBar As Office.CommandBar
    Private selectedBookmark As Microsoft.Office.Tools.Word.Bookmark
    WithEvents boldText As Office.CommandBarButton
    WithEvents ItalicText As Office.CommandBarButton
    '</Snippet4>


    '--------------------------------------------------------------------------
    '<Snippet6>
    Private Sub ThisDocument_Startup(ByVal sender As Object, ByVal e As _
        System.EventArgs) Handles Me.Startup

        AddPopUp()
    End Sub
    '</Snippet6>


    '--------------------------------------------------------------------------
    '<Snippet5>
    Private Sub AddPopUp()

        commandBar = Application.CommandBars.Add( _
            "FormatText", Office.MsoBarPosition.msoBarPopup, , True)

        ' Add a button and set the style, caption, icon and tag.
        boldText = CType(commandBar.Controls.Add(1), Office.CommandBarButton)

        With boldText
            .Style = Office.MsoButtonStyle.msoButtonIconAndCaption
            .Caption = "Bold"
            .FaceId = 113
            .Tag = "0"
        End With

        ' Add a button and set the style, caption, icon and tag.
        ItalicText = CType(commandBar.Controls.Add(1), Office.CommandBarButton)

        With ItalicText
            .Style = Office.MsoButtonStyle.msoButtonIconAndCaption
            .Caption = "Italic"
            .FaceId = 114
            .Tag = "1"
        End With

        CType(Me.AttachedTemplate, Microsoft.Office.Interop.Word.Template).Saved = True

    End Sub
    '</Snippet5>


    '--------------------------------------------------------------------------
     '<Snippet8>
    Private Sub showPopupMenu(ByVal sender As Object, ByVal e _
        As Microsoft.Office.Tools.Word.ClickEventArgs) _
        Handles Bookmark1.BeforeRightClick, Bookmark2.BeforeRightClick

        Dim startPosition As Integer = 0
        Dim i As Integer

        ' If bookmarks overlap, get bookmark closest to cursor.
        For i = 1 To e.Selection.Bookmarks.Count
            If e.Selection.Bookmarks(i).Start > startPosition Then
                startPosition = e.Selection.Bookmarks(i).Start
            End If
        Next

        ' If closest bookmark is the sender, show the popup.
        If DirectCast(sender, Microsoft.Office.Tools.Word.Bookmark).Start = _
            startPosition Then
            selectedBookmark = DirectCast(sender, Microsoft.Office.Tools.Word.Bookmark)
            commandBar.ShowPopup()
            e.Cancel = True
        End If
    End Sub
    '</Snippet8>


    '--------------------------------------------------------------------------
    '<Snippet9>
    Private Sub ButtonClick(ByVal ctrl As Office.CommandBarButton, _
        ByRef Cancel As Boolean) Handles boldText.Click, ItalicText.Click

        Select Case ctrl.Caption
            Case "Bold"
                selectedBookmark.Bold = Not selectedBookmark.Bold
            Case "Italic"
                selectedBookmark.Italic = Not selectedBookmark.Italic
        End Select

        Me.ActiveWindow.SetFocus()
    End Sub
    '</Snippet9>


    '--------------------------------------------------------------------------
    Sub test1()

        '<Snippet10>
        Dim targetControl As Object
        For Each targetControl In Me.Controls

            If TypeOf (targetControl) Is Microsoft.Office.Tools.Word.Bookmark Then
                Dim bookMark As Microsoft.Office.Tools.Word.Bookmark = _
                    CType(targetControl, Microsoft.Office.Tools.Word.Bookmark)

                ' Do some work with the book mark here.
            End If
        Next
        '</Snippet10>


        '<Snippet2>
        Me.SampleBookmark.Start = Me.SampleBookmark.Start - 5
        '</Snippet2>


        '<Snippet3>
        Me.SampleBookmark.End = Me.SampleBookmark.End + 5
        '</Snippet3>


        '<Snippet1>
        Dim firstParagraph As Microsoft.Office.Tools.Word.Bookmark
        firstParagraph = Me.Controls.AddBookmark(Me.Paragraphs(1).Range, "FirstParagraph")
        '</Snippet1>
    End Sub

End Class
